File: audit/f2-kpi-arithmetic.py
#!/usr/bin/env python3
"""F2 — the same arithmetic straight from the committed GeoJSON, no browser."""
import json, os
D = os.path.join(os.path.dirname(__file__), "..", "data", "morocco")
fs = json.load(open(os.path.join(D, "power-plants.geojson")))["features"]
cap = lambda f: f["properties"].get("capacity_mw") or 0
ft  = lambda f: f["properties"].get("fuel_type", "")
st  = lambda f: f["properties"].get("status")
tot      = sum(map(cap, fs))
coded    = sum(cap(f) for f in fs if ft(f) in ("solar", "wind", "hydro"))
correct  = sum(cap(f) for f in fs if ft(f).split("_")[0] in ("solar", "wind", "hydro"))
ops      = [f for f in fs if st(f) == "operational"]
op_tot   = sum(map(cap, ops))
op_ren   = sum(cap(f) for f in ops if ft(f).split("_")[0] in ("solar", "wind", "hydro"))
print(f"fuel_type values          {sorted({ft(f) for f in fs})}")
print(f"features with 'solar'     {sum(1 for f in fs if ft(f)=='solar')}")
print(f"denominator               {tot} MW")
print(f"numerator as coded        {coded} MW -> {round(100*coded/tot)}%   <-- DISPLAYED")
print(f"numerator corrected       {correct} MW -> {round(100*correct/tot)}%")
print(f"operational-only          {op_ren}/{op_tot} -> {round(100*op_ren/op_tot)}%")
print(f"solar MW dropped          {correct-coded} MW")
print(f"pumped_storage counted    {sum(cap(f) for f in fs if ft(f)=='pumped_storage')} MW")
print(f"not-yet-built counted     {tot-op_tot} MW")
assert round(100*coded/tot) == 23 and round(100*correct/tot) == 39
print("\nF2 REPRODUCED")

File: audit/f2-kpi-arithmetic.js
// F2 — "Renewables share" excludes solar_pv/solar_csp because the filter tests
// the literal string "solar", which occurs in zero features.
const {open,show}=require('./_harness');
(async()=>{
  const {page,close}=await open();
  const r=await page.evaluate(()=>{
    const fs=window.__map.getSource('src-power')._data.features;
    const cap=f=>f.properties.capacity_mw||0, ft=f=>f.properties.fuel_type||'';
    const sum=a=>a.reduce((s,f)=>s+cap(f),0);
    const ops=fs.filter(f=>f.properties.status==='operational');
    return {
      displayedKPIs:document.querySelector('#kpiGrid').innerText.replace(/\n/g,' | '),
      fuelTypesInData:[...new Set(fs.map(ft))].sort(),
      matchesLiteralSolar:fs.filter(f=>ft(f)==='solar').length,
      denominatorAllStatuses:sum(fs),
      numeratorAsCoded:sum(fs.filter(f=>['solar','wind','hydro'].includes(ft(f)))),
      numeratorWithSolarVariants:sum(fs.filter(f=>['solar','wind','hydro'].includes(ft(f).split('_')[0]))),
      solarMwDropped:sum(fs.filter(f=>ft(f).startsWith('solar'))),
      pumpedStorageInDenominator:sum(fs.filter(f=>ft(f)==='pumped_storage')),
      nonOperationalInDenominator:sum(fs)-sum(ops),
      shareAsCoded:Math.round(100*sum(fs.filter(f=>['solar','wind','hydro'].includes(ft(f))))/sum(fs)),
      shareCorrected:Math.round(100*sum(fs.filter(f=>['solar','wind','hydro'].includes(ft(f).split('_')[0])))/sum(fs)),
      shareOperationalOnly:Math.round(100*sum(ops.filter(f=>['solar','wind','hydro'].includes(ft(f).split('_')[0])))/sum(ops))
    };
  });
  show(r);
  const ok = r.matchesLiteralSolar===0 && r.shareAsCoded===23 && r.shareCorrected===39
             && /RENEWABLES SHARE\* \| 23%/.test(r.displayedKPIs);
  console.log(ok?'\nF2 REPRODUCED':'\nF2 NOT reproduced');
  await close(); process.exit(ok?0:1);
})();
